### PR TITLE
[Workspace] Make url stateful in hash with workspace id

### LINKED ISSUE #6015: [Workspace] Make url stateful in url with workspace id
**Is your feature request related to a problem? Please describe.**

In order to let users switch to specific workspace when paste a deep link from other users, we need to make url stateful with workspace info. This PR mainly introduce the capability to keep the workspace Id in url consistent with the workspace id in memory so that user can feel free to copy the url and share to others.

**Describe the solution you'd like**

Use the hash to store current workspace id

**Describe alternatives you've considered**

#5243 

**Additional context**

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

In order to let users switch to specific workspace when paste a deep link from other users, we need to make url stateful with workspace info. This PR mainly introduce the capability to keep the workspace Id in url consistent with the workspace id in memory so that user can feel free to copy the url and share to others.

### Issues Resolved

#6015 

## Testing the changes

- Using the branch to bootstrap
- Enable workspace by adding `workspace.enabled` to opensearch_dashboards.yml file 
- Start OSD
- Navigate to any page, append with hash `#/?_w=foo`
- Jump to any other pages by clicking nav on left navigation
- You will find _w params will always be there, making the url shareable.

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by SuZhou-Joe** [comment]
@kavilla Could you please take a look on this PR as you have enough context on both the issue and the related RFC?

**Review by ashwin-pc — COMMENTED** [review]
Hmm this has an issue. It does not hook into the default routing API's of the app. This has some risks, e.g. opening links in a new tab to name just the first that comes to mind. Take a look at Discover. Go to discover, expand a row, click on view surrounding documents. This will not work there since the workspace is not carried over. This is just one place. Have you looked at https://ashwin-pc.github.io/OpenSearch-Dashboards/docs/index.html#/plugins/data_persistence?

Also this should definitely have a test. That way we can ensure that under all circumstances the workspace key is passed :)

**Comment by SuZhou-Joe** [comment]
@ashwin-pc So grateful for the edge case.
- For the edge case in Discover, I add a localStorage to reserve the current workspace so that the new-tab-link will have workspace params patched without doing any code changes. And the approach should solve the concern for breadcrumbs and hard refresh when navigating inside OSD.
- For the test cases, sure I will add cypress test to cover the edge cases.

File: src/plugins/workspace/public/plugin.ts [hunk 1]
```diff
@@ -3,16 +3,110 @@
...
+
+    return (await this.core?.workspaces.currentWorkspaceId$.getValue()) || '';
+  }
+  private getPatchedUrl = (url: string, workspaceId: string) => {
+    return formatUrlWithWorkspaceId(url, workspaceId);
+  };
+  private hashChangeHandler = async () => {
+    if (this.shouldPatchUrl()) {
+      const workspaceId = await this.getWorkpsaceId();
+      this.URLChange$.next(this.getPatchedUrl(window.location.href, workspaceId));
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.ts:60]: Yes, in that case we will remove the _w params from url hash.

**Comment by SuZhou-Joe** [comment]
@ashwin-pc , we had a discussion on all the options we can choose from. This is the RFC issue(https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5243).

@kavilla has some concerns on making the workspaceId into path as it may affect the OSD a lot, so we did a trade-off to make it in hash. 

I am still free to accept advice on how we store the workspace id into url and look forward to your advice as you have much more knowledge on OSD router.

**Comment by SuZhou-Joe** [comment]
Closed as we will use #6060.
